README.md
rtel
====

Lightweight TerminalEmulator

ChangeLog
2009-12-08 12:39  xbatob

	* rtel.spec, src/tty.c: в случае ошибки работы с tty пропущен один
	  return

2009-11-30 13:57  xbatob

	* rtel.kdevelop, rtel.spec, src/main.c: fixes for new libc

2009-02-18 14:29  xbatob

	* rtel.kdevelop, rtel.spec, src/main.c: исправлена фатальная ошибка

2009-02-18 11:37  xbatob

	* rtel.spec, src/main.c, src/tty.c, src/tty.h: Исправлена установка
	  паритетаб добавлена установка размера символа

2009-02-12 07:47  xbatob

	* configure.in, po/rtel.pot, po/ru.po, rtel.kdevelop, rtel.spec,
	  src/main.c: - выкинут popt;
	  - добавлена возможность установки паритета на линии

2008-12-09 12:47  xbatob

	* config.h.in: очередные проблемы с установкой - не клался
	  ld-linux.so, не все нужные файлы копировались в /etc.
	  Опять - ненужные файлы

2008-08-13 13:46  xbatob

	* rtel.kdevelop, rtel.spec, src/main.c: был пропущеи break в
	  разборе параметров

2008-06-24 10:20  xbatob

	* po/rtel.pot, po/ru.po: обновлён перевод

2008-06-24 10:09  xbatob

	* ChangeLog, config.h.in, configure.in, rtel.kdevelop, rtel.spec,
	  src/main.c: Может быть собрана с использованием popt или argp для
	  разбора коммандной строки (у обоих методов есть недостатки).

2008-06-16 11:22  xbatob

	* FILES: filelist fixed

2008-06-16 11:12  xbatob

	* rtel.kdevelop, src/main.c: conflicting option removed

2008-06-16 11:08  xbatob

	* rtel.kdevelop, src/Makefile.am, src/debug.h, src/node.c,
	  src/rtel.c: debug code added

2008-06-16 09:48  xbatob

	* src/rtel.c: #include missed

2008-06-16 09:45  xbatob

	* rtel.kdevelop, src/rtel.c: bug fixed - doesn't compile w/o
	  WITH_FILTER

2008-06-16 09:32  xbatob

	* aclocal.m4, autom4te.cache: extra files deleted

2008-06-16 09:27  xbatob

	* ABOUT-NLS, FILES, Makefile.in, aclocal.m4,
	  autom4te.cache/output.1, autom4te.cache/traces.1, configure,
	  configure.in, po/rtel.pot, po/ru.po, rtel.kdevelop, rtel.spec,
	  src/Makefile.am, src/main.c, src/node.c, src/node.h,
	  src/ringbuf.c, src/ringbuf.h, src/rtel.c, src/rtel.h,
	  templates/c, templates/h: filters implemented

2008-06-07 10:40  xbatob

	* Makefile.in, rtel.kdevelop, src/main.c: ident fixed

2008-06-03 18:27  xbatob

	* Makefile.in, autom4te.cache/requests, po/rtel.pot, po/ru.po:
	  translation updated

2008-06-03 10:47  xbatob

	* autom4te.cache/output.1, autom4te.cache/traces.1, config.h.in,
	  configure, configure.in, rtel.kdevelop, rtel.spec, src/main.c,
	  src/rtel.c, src/rtel.h: Can accept name of logfile via readline()

2008-05-14 14:05  xbatob

	* FILES, autom4te.cache/requests, po/rtel.pot, po/ru.po, rtel.spec,
	  src/Makefile.am, src/main.c, src/msgs.h, src/rtel.c, src/tty.c:
	  diagnostic improved

2008-04-29 10:21  xbatob

	* po/rtel.pot, po/ru.po: translations updated

2008-04-29 09:25  xbatob

	* rtel.kdevelop, rtel.spec, src/main.c: Fixed help & diagnostic
	  messages.
	  Translations are NOT UPDATED.

2008-04-28 08:33  xbatob

	* src/main.c: version string is formatted

2008-04-28 08:28  xbatob

	* src/gettext.h, src/main.c, src/msgs.h, src/ringbuf.c,
	  src/ringbuf.h, src/rtel.c, src/rtel.h, src/rts.h, src/tty.c,
	  src/tty.h: version string is more informative

2008-04-28 08:20  xbatob

	* src/main.c: version string is more informative

2008-04-28 07:51  xbatob

	* rtel.spec, src/main.c: Can leave device speed unchanged

2008-04-23 08:37  xbatob

	* rtel.kdevelop, src/Makefile.am: redundant linking with libpopt
	  removed

2008-04-23 08:34  xbatob

	* Makefile.in, aclocal.m4, autom4te.cache/requests,
	  autom4te.cache/traces.1, rtel.kdevelop, src/Makefile.in,
	  src/main.c: deleted unneeded files

2008-04-01 11:04  xbatob

	* po/rtel.pot, po/ru.po: translations updated

2008-04-01 10:49  xbatob

	* Makefile.in, aclocal.m4, autom4te.cache/output.1,
	  autom4te.cache/requests, autom4te.cache/traces.1, configure,
	  configure.in, rtel.kdevelop, rtel.spec, src/Makefile.in,
	  src/main.c, src/rtel.c, src/rtel.h: can log input

2008-03-28 11:40  xbatob

	* po/ru.po, rtel.kdevelop: translation fixed :-)

2008-03-27 12:00  xbatob

	* po/ru.po, rtel.kdevelop: translation fixed :-)

2008-03-24 13:50  xbatob

	* autom4te.cache/output.1, autom4te.cache/requests,
	  autom4te.cache/traces.1, config.h.in, configure, configure.in,
	  po/rtel.pot, po/ru.po, rtel.kdevelop, rtel.spec, src/Makefile.am,
	  src/Makefile.in, src/main.c, src/tty.c, src/tty.h: use popt
	  instead of getopt

2008-03-11 13:27  xbatob

	* src/gettext.h: localisation + russuan messages

2008-03-11 13:22  xbatob

	* ChangeLog, FILES, Makefile.am, Makefile.in, aclocal.m4,
	  autom4te.cache/output.1, autom4te.cache/output.2,
	  autom4te.cache/requests, autom4te.cache/traces.1, config.h.in,
	  config.rpath, configure, configure.in, m4, m4/ChangeLog,
	  m4/gettext.m4, m4/iconv.m4, m4/lib-ld.m4, m4/lib-link.m4,
	  m4/lib-prefix.m4, m4/nls.m4, m4/po.m4, m4/progtest.m4, po,
	  po/ChangeLog, po/LINGUAS, po/Makefile.in.in, po/Makevars,
	  po/POTFILES.in, po/Rules-quot, po/boldquot.sed,
	  po/en@boldquot.header, po/en@quot.header, po/insert-header.sin,
	  po/quot.sed, po/remove-potcdate.sin, po/rtel.pot, po/ru.po,
	  rtel.kdevelop, rtel.spec, src/Makefile.am, src/Makefile.in,
	  src/main.c, src/msgs.h, src/rtel.c: localisation + russuan
	  messages

2008-02-20 10:51  xbatob

	* src/main.c: cosmetics

2008-02-20 10:44  xbatob

	* src/main.c: info string changed

2008-02-20 10:35  xbatob

	* aclocal.m4, autom4te.cache/output.1, autom4te.cache/output.2,
	  autom4te.cache/requests, autom4te.cache/traces.1,
	  autom4te.cache/traces.2, configure, configure.in, rtel.spec,
	  src/Makefile.in, src/main.c, src/ringbuf.c, src/ringbuf.h,
	  src/rtel.c, src/rtel.h: - can change escape character
	  - online help implemented

2007-11-13 13:29  maxs

	* src/tty.h: Just a test of SVN functionality under windows

2007-10-27 08:34  xbatob

	* autom4te.cache/requests, configure.in, rtel.kdevelop, rtel.spec:
	  package version fixed

2007-10-17 11:43  xbatob

	* autom4te.cache/requests, rtel.kdevelop, src/main.c, src/msgs.h,

src/main.c
/*
*  C Implementation: main
*
* Description:
*
*
* Author: Serguei B. Khvatov <xbatob@techno.spb.ru>, (C) 2005
*
* Copyright: See COPYING file that comes with this distribution
*
* $Id: main.c 2212 2010-06-24 13:18:03Z xbatob $
*/

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include "tty.h"
#include "rts.h"
#include "rtel.h"
#include "gettext.h"

#ifndef LOCALEDIR
#define LOCALEDIR NULL
#endif

#include <sys/socket.h>
#include <netinet/in.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <locale.h>
#include <error.h>
#include <errno.h>

#ifndef PACKAGE_STRING
#define PACKAGE_STRING "lighthweight serial communicator 1.5"
#define PACKAGE_BUGREPORT "Sergey Khvatov <xbatob@techno.spb.ru>"
#endif


TtyState oldopts;

struct rts rts;
static const char *dev;
static int devd;
static speed_t speed = B0;
static tcflag_t par_clr = CRTSCTS, par_set = 0;

#include <argp.h>

const char *argp_program_version = PACKAGE_STRING " ("
#ifdef  HAVE_LIBREADLINE
    "READLINE,"
#endif
#ifdef  WITH_FILTER
    "FILTER,"
#endif
#ifdef  WITH_LOGGING
    "LOGGING,"
#endif
    ")\n$Id: main.c 2212 2010-06-24 13:18:03Z xbatob $";
const char *argp_program_bug_address = PACKAGE_BUGREPORT;

static int init_hook(struct termios *state) {
    state->c_cflag &= ~par_clr;
    state->c_cflag |= par_set;
    return 0;
}

static error_t apf (int key, char *arg, struct argp_state *state) {
    switch (key) {
    case 'd':
    case ARGP_KEY_ARG:
        dev = arg;
        break;
    case 's':
        speed = ttyParseSpeed(strtoul (arg, 0, 10));
        if (speed == B0)
            argp_error (state, _("bad speed value: %s"), arg);
        break;
    case 'o': {
        static const char *pars[] = {
            "odd", "even", "none", "cs8", "cs7",
            0
        };
#define STTY_PARS "odd,even,none,cs8,cs7"
        char *val;
        while (*arg) {
            int i = getsubopt (&arg, (char *const *)pars, &val);
            if (i < 0)
                argp_error (state, _("bad tty option: %s"), val);
            switch (i) {
            case 0:
                par_set |= PARENB | PARODD;
                break;
            case 1:
                par_set = (par_set & ~PARODD) | PARENB;
                par_clr |= PARODD;
                break;
            case 2:
                par_set &= ~(PARENB | PARODD);
                par_clr |= PARENB | PARODD;
                break;
            case 3:
                par_clr |= CSIZE;
                par_set = (par_set & ~CSIZE) | CS8;
                break;
            case 4:
                par_clr |= CSIZE;
                par_set = (par_set & ~CSIZE) | CS7;
                break;
            }
        }
        break;
    }
    case 'r':
        if (!parse_rts (arg))
            argp_error (state, _("bad RTS control value: %s"), arg);
        break;
    case 'e':
        escape = *arg;
        break;
    case ARGP_KEY_END:
        if (!dev)
            argp_error (state, _("device is not set"));
        devd = ttyInit (dev, speed, init_hook);
        if (devd < 0)
            argp_failure (state, 1, 0, _("can not init tty"));
        break;
    default:
        return  ARGP_ERR_UNKNOWN;
    }
    return 0;
}

static char *aph (int key, const char *text, void *input) {
    switch (key) {
    case 'e': {
        char *b;
        asprintf(&b, text, escape);
        return b;
    }
    case 'o': {
        char *b;
        asprintf(&b, text, STTY_PARS);
        return b;
    }
    }
    return (char *)text;
}

static const struct argp_option apo[] = {
    {
        "device", 'd', "FILE", 0,
        N_("Device name")
    }, {
        "speed", 's', "SPEED", 0,
        N_("Set device speed")
    }, {
        "option", 'o', "LIST", 0,
        N_("Coma-separated list of tty options to set (%s)")
    }, {
        "rts", 'r', "1|0|t|r", 0,
        N_("Set rts line to given state"
        " (1 or 0 allways, 1 on transmit, 1 on receive)")
    }, {"escape",  'e', "CHAR", 0,
        N_("Escape character [%c]")
    }, {
        0
    }
};

static const struct argp ap = {
    apo, apf, N_("[device]"),
    N_("lighthweight serial communicator"),
    0, aph
};


int main(int argc, char *argv[]) {
#if			ENABLE_NLS
    setlocale (LC_ALL, "");
    bindtextdomain (PACKAGE, LOCALEDIR);
    textdomain (PACKAGE);
#endif	//	ENABLE_NLS

    argp_parse (&ap, argc, argv, 0, 0, 0);
    rt_help();
    if (!ttySetRaw (0, &oldopts))
        return EXIT_FAILURE;
    rtel_loop (devd);
    ttyReset (0, &oldopts);
    close (devd);
    error (0, 0, _("\nGood Bye!"));
    return EXIT_SUCCESS;
}
